File: 0101-0200/0106-Construct Binary Tree from Inorder and Postorder Traversal/0106-Construct Binary Tree from Inorder and Postorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        if (postorder.size() == 0) {
            return nullptr;
        }
        for (int i = 0; i < inorder.size(); ++i) {
            table[inorder[i]] = i;
        }
        int postIndex = postorder.size() - 1;
        return dfs(inorder, 0, inorder.size() - 1, postorder, postIndex);
    }
    
private:
    unordered_map<int, int> table;
    TreeNode* dfs(vector<int>& inorder, int inStart, int inEnd, vector<int>& postorder, int& postIndex) {
        if (inStart > inEnd) {
            return nullptr;
        }
        
        TreeNode* root = new TreeNode(postorder[postIndex--]);
        int index = table[root->val];
        
        root->right = dfs(inorder, index + 1, inEnd, postorder, postIndex);
        root->left = dfs(inorder, inStart, index - 1, postorder, postIndex);
        return root;
    }
};

File: 0101-0200/0106-Construct Binary Tree from Inorder and Postorder Traversal/0106-Construct Binary Tree from Inorder and Postorder Traversal.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> TreeNode:
        table = {value:index for index, value in enumerate(inorder)}
        postIndex = len(postorder) - 1
        def dfs(inStart, inEnd):
            nonlocal postIndex
            if inStart > inEnd:
                return None
            root = TreeNode(postorder[postIndex])
            postIndex -= 1
            index = table[root.val]
            root.right = dfs(index + 1, inEnd)
            root.left = dfs(inStart, index - 1)
            return root
        return dfs(0, len(inorder) - 1)
